### PR TITLE
NVMEDevice: Fix the issue of multiple OSDs when using SPDK

### PR DESCRIPTION
**Pull request by optimistyzy**
For each OSD when using SPDK, it should be the type of primary
process. So when there are multiple NVMe SSDs, each OSD will use
one. So we do not want an OSD process claim all the NVMe SSDs,
so we use the allowed list to restrict the probe.

And with this patch, we can support multiple OSDs while using
SPDK.

Signed-off-by: Ziye Yang <ziye.yang@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/blk/spdk/NVMEDevice.cc [hunk 1]
```diff
@@ -569,15 +575,18 @@ int NVMEManager::try_get(const spdk_nvme_transport_id& trid, SharedDriverData **
 
   if (!dpdk_thread.joinable()) {
     dpdk_thread = std::thread(
-      [this, coremask_arg, m_core_arg, mem_size_arg]() {
+      [this, coremask_arg, m_core_arg, mem_size_arg, pci_addr]() {
         static struct spdk_env_opts opts;
+        static struct spdk_pci_addr addr = pci_addr;
```
Comment (tchaikov) [inline src/blk/spdk/NVMEDevice.cc:580]: why do we need to use a static variable for `spdk_env_opts` and `spdk_pci_addr`? i think `spdk_env_init()` is a synchronized function, in the sense that we are not obliged to keep its parameter alive after it returns.

File: src/blk/spdk/NVMEDevice.cc [hunk 2]
```diff
@@ -555,6 +555,12 @@ int NVMEManager::try_get(const spdk_nvme_transport_id& trid, SharedDriverData **
     }
   }
 
+  struct spdk_pci_addr pci_addr;
+  int rc = spdk_pci_addr_parse(&pci_addr, trid.traddr);
+  if (rc < 0) {
+    derr << __func__ << " invalid transport address" << trid.traddr << dendl;
```
Comment (tchaikov) [inline src/blk/spdk/NVMEDevice.cc:561]:
```suggestion
    derr << __func__ << " invalid transport address: " << trid.traddr << dendl;
```

File: src/blk/spdk/NVMEDevice.cc (hunk 1 above)
Comment (optimistyzy) [reply src/blk/spdk/NVMEDevice.cc:580]: pci_addr is const, so we use another variable, we can change the type, and remove the static.

File: src/blk/spdk/NVMEDevice.cc (hunk 2 above)
Comment (optimistyzy) [reply src/blk/spdk/NVMEDevice.cc:561]: ack

**Comment by optimistyzy** [comment]
Added @majianpeng. He finds this issue in kernel 5.4, and I also find this issue in kernel 5.7. With this patch, it can solve the issues in both.

**Comment by optimistyzy** [comment]
Hi @tchaikov , I see the make check test failed, but it seems not related with this PR.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-07-10_14:43:26-rados-wip-kefu-testing-2020-07-10-1523-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-11_05:07:55-rados-wip-kefu-testing-2020-07-11-1109-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-11_08:56:15-rados-wip-kefu-testing-2020-07-11-1458-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/46442

**Pull request merged**